core/src/banking_stage.rs

```diff
@@ -1077,6 +1015,300 @@ impl BankingStage {
         }
     }
 
+    /// Filter out packets that fail to sanitize, or are no longer valid (could be
+    /// too old, a duplicate of something already processed). Doing this in batches to avoid
+    /// checking bank's blockhash and status cache per transaction which could be bad for performance.
+    /// Added valid and sanitized packets to forwarding queue.
+    pub fn filter_and_forward_with_account_limits(
+        bank: &Arc<Bank>,
+        buffered_packet_batches: &mut UnprocessedPacketBatches,
+        forward_buffer: &mut ForwardPacketBatchesByAccounts,
+        batch_size: usize,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+        banking_stage_stats: &BankingStageStats,
+    ) -> FilterForwardingResults {
+        let mut total_forwardable_tracer_packets: usize = 0;
+        let mut total_tracer_packets_in_buffer: usize = 0;
+        let mut total_forwardable_packets: usize = 0;
+        let mut dropped_tx_before_forwarding_count: usize = 0;
+
+        let mut original_priority_queue = Self::swap_priority_queue(buffered_packet_batches);
+
+        // indicates if `forward_buffer` still accept more packets, see details at
+        // `ForwardPacketBatchesByAccounts.rs`.
+        let mut accepting_packets = true;
+        // batch iterate through buffered_packet_batches in desc priority order
+        let retained_priority_queue: MinMaxHeap<Rc<ImmutableDeserializedPacket>> =
+            original_priority_queue
+                .drain_desc()
+                .chunks(batch_size)
+                .into_iter()
+                .flat_map(|packets_to_process| {
+                    let packets_to_process = packets_to_process.into_iter().collect_vec();
```

Comment:
If we didn't run the chunks and evaluated each packet independently, I think we could avoid cloning/allocating this vector

Reply:
do it transaction by transaction would significantly increase lock contention in bank, from description: "When checking if packet is invalid, should keep in mind not to lock bank's blockhash_queue and status_cache for every transactions, as that's be bad for overall performance. This requires to change the forwarder's workflow from transaction-by-transaction to batch-by-batch."

Reply:
got it, yeah we'll get around to optimizing that blockhash queue at some point...